src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AvoidInfiniteRecursion.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class AvoidInfiniteRecursion : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2008";
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+            new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AvoidInfiniteRecursionTitle), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources)),
+            new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AvoidInfiniteRecursionMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources)),
+            DiagnosticCategory.Reliability,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            helpLinkUri: null);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationBlockStartAction(operationBlockStartContext =>
+            {
+                if (operationBlockStartContext.OwningSymbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.MethodKind == MethodKind.PropertySet)
+                {
+                    operationBlockStartContext.RegisterOperationAction(operationContext =>
+                    {
+                        var assignmentOperation = (IAssignmentOperation)operationContext.Operation;
+
+                        if (assignmentOperation.Target is IPropertyReferenceOperation operationTarget &&
+                            operationTarget.Member.Equals(methodSymbol.AssociatedSymbol))
+                        {
+                            operationContext.ReportDiagnostic(Diagnostic.Create(Rule, assignmentOperation.Syntax.GetLocation(), operationTarget.Property.Name));
```

Comment:
Implementation looks good to me, but on second thought I am wondering if you need to handle cases where the assignment happens inside a lambda or local function within the setter. That does seem like a contrived case, but we can choose to be more defensive by making sure that we have an additional check here that none of the ancestors of this assignment operation are `IAnonymousFunctionOperation` or `ILocalFunctionOperation`. I would let you make a call on this.

Reply:
Good point! I tend to think that's better to exclude the cases inside lambdas as we won't precisely know how it will be used.